src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,272 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+import util.Utility;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+
+public class LabelPickerState {
+    Set<String> addedLabels;
+    Set<String> removedLabels;
+    Set<String> initialLabels;
+    List<String> matchedLabels;
+    OptionalInt currentSuggestionIndex;
+
+    private LabelPickerState(Set<String> initialLabels,
+                             Set<String> addedLabels,
+                             Set<String> removedLabels,
+                             List<String> matchedLabels,
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             Set<String> addedLabels,
+                             Set<String> removedLabels,
+                             List<String> matchedLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    /*
+     * Methods for Logic
+     */
+
+    public LabelPickerState(Set<String> initialLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = new HashSet<>();
+        this.removedLabels = new HashSet<>();
+        this.matchedLabels = new ArrayList<>();
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    public LabelPickerState clearMatchedLabels() {
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels,
+                    new ArrayList<>(), currentSuggestionIndex);
+    }
+
+    public LabelPickerState toggleLabel(Set<String> repoLabels, String keyword) {
+        if (!hasExactlyOneMatchedLabel(repoLabels, keyword)) return this;
+        String labelName = getMatchedLabelName(repoLabels, keyword);
+
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                // add back initial label
+                removeConflictingLabels(labelName);
+                removedLabels.remove(labelName);
```

Comment:
Hmm, so this class isn't immutable. I'm not sure if this is a problem, as I know it's a huge hassle to build up the intermediate collections.